lms/djangoapps/grades/models.py

```diff
@@ -153,31 +153,33 @@ def blocks(self):
         return BlockRecordList.from_json(self.blocks_json)
 
     @classmethod
-    def bulk_read(cls, course_key):
+    def bulk_read(cls, user_id, course_key):
         """
-        Reads and returns all visible block records for the given course from
-        the cache.  The cache is initialize with the visible blocks for this
-        course if no entry currently exists.has no entry for this course,
-        the cache is updated.
+        Reads and returns all visible block records for the given user and course from
+        the cache.  The cache is initialized with the visible blocks for this user and
+        course if no entry currently exists.
 
         Arguments:
             course_key: The course identifier for the desired records
         """
-        prefetched = get_cache(cls._CACHE_NAMESPACE).get(cls._cache_key(course_key), None)
+        prefetched = get_cache(cls._CACHE_NAMESPACE).get(cls._cache_key(user_id, course_key), None)
         if prefetched is None:
-            prefetched = cls._initialize_cache(course_key)
+            prefetched = cls._initialize_cache(user_id, course_key)
         return prefetched
 
     @classmethod
-    def cached_get_or_create(cls, blocks):
-        prefetched = get_cache(cls._CACHE_NAMESPACE).get(cls._cache_key(blocks.course_key))
+    def cached_get_or_create(cls, user_id, blocks):
+        prefetched = get_cache(cls._CACHE_NAMESPACE).get(cls._cache_key(user_id, blocks.course_key))
         if prefetched is not None:
             model = prefetched.get(blocks.hash_value)
             if not model:
-                model = cls.objects.create(
+                # We still have to do a get_or_create, because
+                # another user may have had this block hash created,
+                # even if the user we checked the cache for hasn't yet.
```

Comment:
hmm.. interesting.  I can see this happening with our [repeatable-read isolation level](https://stackoverflow.com/a/4036063) setting on MySQL.   Good catch.

Reply:
I actually caught this due to Unique constraint violation during unit testing.